README.md
[![Github Build](https://github.com/scriptex/gitlab-calendar/workflows/Build/badge.svg)](https://github.com/scriptex/gitlab-calendar/actions?query=workflow%3ABuild)
[![Codacy Badge](https://app.codacy.com/project/badge/Grade/34d3d75710534dc6a38c3584a1dcd068)](https://www.codacy.com/gh/scriptex/gitlab-calendar/dashboard?utm_source=github.com&utm_medium=referral&utm_content=scriptex/gitlab-calendar&utm_campaign=Badge_Grade)
[![Codebeat Badge](https://codebeat.co/badges/d765a4c8-2c0e-44f2-89c3-fa364fdc14e6)](https://codebeat.co/projects/github-com-scriptex-gitlab-calendar-master)
[![CodeFactor Badge](https://www.codefactor.io/repository/github/scriptex/gitlab-calendar/badge)](https://www.codefactor.io/repository/github/scriptex/gitlab-calendar)
[![DeepScan grade](https://deepscan.io/api/teams/3574/projects/5257/branches/40799/badge/grade.svg)](https://deepscan.io/dashboard#view=project&tid=3574&pid=5257&bid=40799)
[![Analytics](https://ga-beacon-361907.ew.r.appspot.com/UA-83446952-1/github.com/scriptex/gitlab-calendar/README.md?pixel)](https://github.com/scriptex/gitlab-calendar/)

# Gitlab Calendar

> Embed your Gitlab Contributions Calendar anywhere

This package allows you to embed your Gitlab Contributions Calendar in any website.

The package does not handle network requests, this is left to the developer to decide and implement. Gitlab Calendar requires the data as it is provided by the `calendar` endpoint of the Gitlab API (https://gitlab.com/users/$USER/calendar.json).

**This project was inspired by [Github Calendar](https://github.com/Bloggify/github-calendar).**

## Visitor stats

![GitHub stars](https://img.shields.io/github/stars/scriptex/gitlab-calendar?style=social)
![GitHub forks](https://img.shields.io/github/forks/scriptex/gitlab-calendar?style=social)
![GitHub watchers](https://img.shields.io/github/watchers/scriptex/gitlab-calendar?style=social)
![GitHub followers](https://img.shields.io/github/followers/scriptex?style=social)

## Code stats

![GitHub code size in bytes](https://img.shields.io/github/languages/code-size/scriptex/gitlab-calendar)
![GitHub repo size](https://img.shields.io/github/repo-size/scriptex/gitlab-calendar?style=plastic)
![GitHub language count](https://img.shields.io/github/languages/count/scriptex/gitlab-calendar?style=plastic)
![GitHub top language](https://img.shields.io/github/languages/top/scriptex/gitlab-calendar?style=plastic)
![GitHub last commit](https://img.shields.io/github/last-commit/scriptex/gitlab-calendar?style=plastic)

## Live preview ([link](https://atanas.info/stats))

![Gitlab Calendar Screenshot](https://raw.githubusercontent.com/scriptex/gitlab-calendar/master/assets/gitlab-calendar.png)

## Installation

```sh
npm i gitlab-calendar

# or

yarn add gitlab-calendar
```

## Usage

```typescript
import GitlabCalendar from 'gitlab-calendar';

new GitlabCalendar(document.getElementById('gitlab-calendar'), data, options);
```

## Data format

The data is obtained by fetching the `https://gitlab.com/users/$USER/calendar.json` endpoint.

This exercise is left to the developer implementing it.

`gitlab-calendar` requires the data in the following format:

```json
{
    "2020-12-07": 1,
    "2020-12-09": 2,
    "2020-12-10": 5,
    ...
}
```

## Options

| Name                | Type                                                           | Default value                                                                                                                                                                                                                                                                                              |
| ------------------- | -------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `daySize`           | `number`                                                       | 15                                                                                                                                                                                                                                                                                                         |
| `hintText`          | `string`                                                       | 'Issues, merge requests, pushes, and comments.'                                                                                                                                                                                                                                                            |
| `daySpace`          | `number`                                                       | 1                                                                                                                                                                                                                                                                                                          |
| `utcOffset`         | `number`                                                       | 0                                                                                                                                                                                                                                                                                                          |
| `dayTitles`         | `Record<ActivityCalendarWeekday, string>`                      | <pre lang="json">{ <br> monday: 'M', <br> wednesday: 'W', <br> friday: 'F', <br> saturday: 'S', <br> sunday: 'S' <br>}</pre>                                                                                                                                                                               |
| `monthsAgo`         | `number`                                                       | 12                                                                                                                                                                                                                                                                                                         |
| `monthNames`        | `string[]`                                                     | ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']                                                                                                                                                                                                                       |
| `inputFormat`       | `string`                                                       | 'yyyy-MM-dd'                                                                                                                                                                                                                                                                                               |
| `weekdayNames`      | `string[]`                                                     | ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday']                                                                                                                                                                                                                             |
| `legendValues`      | `ActivityCalendarLegendValue[]`                                | <pre lang="json">[ <br> { title: 'No contributions', min: 0 }, <br> { title: '1-9 contributions', min: 1 }, <br> { title: '10-19 contributions', min: 10 }, <br> { title: '20-29 contributions', min: 20 }, <br> { title: '30+ contributions', min: 30 } <br>]</pre>                                       |
| `firstDayOfWeek`    | `number`                                                       | 0                                                                                                                                                                                                                                                                                                          |
| `tooltipDateFormat` | `string`                                                       | 'MMM d, yyyy'                                                                                                                                                                                                                                                                                              |
| `tooltipFormatter`  | `(count: number, dayName: string, dateText: string) => string` | <pre lang="json">(count: number, dayName: string, dateText: string) => { <br />let contribText = 'No contributions';<br /><br />if (count > 0) {<br />contribText = count === 1 ? '1 contribution' : `${count} contributions`;<br /><br />}return `${contribText} on ${dayName} ${dateText}`;<br />}</pre> |

## LICENSE

MIT

---

<div align="center">
    Connect with me:
</div>

<br />

<div align="center">
    <a href="https://atanas.info">
        <img src="https://raw.githubusercontent.com/scriptex/socials/master/styled-assets/logo.svg" height="20" alt="">
    </a>
    &nbsp;
    <a href="mailto:hi@atanas.info">
        <img src="https://raw.githubusercontent.com/scriptex/socials/master/styled-assets/email.svg" height="20" alt="">
    </a>
    &nbsp;
    <a href="https://www.linkedin.com/in/scriptex/">
        <img src="https://raw.githubusercontent.com/scriptex/socials/master/styled-assets/linkedin.svg" height="20" alt="">
    </a>
    &nbsp;
    <a href="https://github.com/scriptex">
        <img src="https://raw.githubusercontent.com/scriptex/socials/master/styled-assets/github.svg" height="20" alt="">
    </a>
    &nbsp;
    <a href="https://gitlab.com/scriptex">
        <img src="https://raw.githubusercontent.com/scriptex/socials/master/styled-assets/gitlab.svg" height="20" alt="">
    </a>
    &nbsp;
    <a href="https://twitter.com/scriptexbg">
        <img src="https://raw.githubusercontent.com/scriptex/socials/master/styled-assets/twitter.svg" height="20" alt="">
    </a>
    &nbsp;
    <a href="https://www.npmjs.com/~scriptex">
        <img src="https://raw.githubusercontent.com/scriptex/socials/master/styled-assets/npm.svg" height="20" alt="">
    </a>
    &nbsp;
    <a href="https://www.youtube.com/user/scriptex">
        <img src="https://raw.githubusercontent.com/scriptex/socials/master/styled-assets/youtube.svg" height="20" alt="">
    </a>
    &nbsp;
    <a href="https://stackoverflow.com/users/4140082/atanas-atanasov">
        <img src="https://raw.githubusercontent.com/scriptex/socials/master/styled-assets/stackoverflow.svg" height="20" alt="">
    </a>
    &nbsp;
    <a href="https://codepen.io/scriptex/">
        <img src="https://raw.githubusercontent.com/scriptex/socials/master/styled-assets/codepen.svg" width="20" alt="">
    </a>
    &nbsp;
    <a href="https://profile.codersrank.io/user/scriptex">
        <img src="https://raw.githubusercontent.com/scriptex/socials/master/styled-assets/codersrank.svg" height="20" alt="">
    </a>
    &nbsp;
    <a href="https://linktr.ee/scriptex">
        <img src="https://raw.githubusercontent.com/scriptex/socials/master/styled-assets/linktree.svg" height="20" alt="">
    </a>
</div>

---

<div align="center">
Support and sponsor my work:
<br />
<br />
<a href="https://twitter.com/intent/tweet?text=Checkout%20this%20awesome%20developer%20profile%3A&url=https%3A%2F%2Fgithub.com%2Fscriptex&via=scriptexbg&hashtags=software%2Cgithub%2Ccode%2Cawesome" title="Tweet">
	<img src="https://img.shields.io/badge/Tweet-Share_my_profile-blue.svg?logo=twitter&color=38A1F3" />
</a>
<a href="https://paypal.me/scriptex" title="Donate on Paypal">
	<img src="https://img.shields.io/badge/Donate-Support_me_on_PayPal-blue.svg?logo=paypal&color=222d65" />
</a>
<a href="https://revolut.me/scriptex" title="Donate on Revolut">
	<img src="https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/scriptex/scriptex/master/badges/revolut.json" />
</a>
<a href="https://patreon.com/atanas" title="Become a Patron">
	<img src="https://img.shields.io/badge/Become_Patron-Support_me_on_Patreon-blue.svg?logo=patreon&color=e64413" />
</a>
<a href="https://ko-fi.com/scriptex" title="Buy Me A Coffee">
	<img src="https://img.shields.io/badge/Donate-Buy%20me%20a%20coffee-yellow.svg?logo=ko-fi" />
</a>
<a href="https://liberapay.com/scriptex/donate" title="Donate on Liberapay">
	<img src="https://img.shields.io/liberapay/receives/scriptex?label=Donate%20on%20Liberapay&logo=liberapay" />
</a>

<a href="https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/scriptex/scriptex/master/badges/bitcoin.json" title="Donate Bitcoin">
	<img src="https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/scriptex/scriptex/master/badges/bitcoin.json" />
</a>
<a href="https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/scriptex/scriptex/master/badges/etherium.json" title="Donate Etherium">
	<img src="https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/scriptex/scriptex/master/badges/etherium.json" />
</a>
<a href="https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/scriptex/scriptex/master/badges/shiba-inu.json" title="Donate Shiba Inu">
	<img src="https://img.shields.io/endpoint?url=https://raw.githubusercontent.com/scriptex/scriptex/master/badges/shiba-inu.json" />
</a>
</div>


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [2.0.1](https://github.com/scriptex/gitlab-calendar/compare/2.0.0...2.0.1)

> 7 November 2022

- Fix the URL to the demo [`d69f1a1`](https://github.com/scriptex/gitlab-calendar/commit/d69f1a16de0cc1a0931f77d9d43ebe465ac964be)
- Update dependency release-it to v15.4.3 [`f5397c5`](https://github.com/scriptex/gitlab-calendar/commit/f5397c534cf64529a8184bb42bdfd7c15d34610c)
- Update dependency release-it to v15.4.2 [`95f0e8d`](https://github.com/scriptex/gitlab-calendar/commit/95f0e8d74a90957f524f480227fa53d14a0cbb1a)

### [2.0.0](https://github.com/scriptex/gitlab-calendar/compare/1.0.0...2.0.0)

> 12 September 2022

- Update implementation and metadata [`#17`](https://github.com/scriptex/gitlab-calendar/pull/17)
- Bump parse-url from 6.0.0 to 6.0.2 [`#16`](https://github.com/scriptex/gitlab-calendar/pull/16)
- Update dependency release-it to v15 [`#15`](https://github.com/scriptex/gitlab-calendar/pull/15)
- Bump minimist from 1.2.5 to 1.2.6 [`#14`](https://github.com/scriptex/gitlab-calendar/pull/14)
- Update actions/checkout action to v3 [`#11`](https://github.com/scriptex/gitlab-calendar/pull/11)
- Update actions/setup-node action to v3 [`#10`](https://github.com/scriptex/gitlab-calendar/pull/10)
- Bump node-fetch from 2.6.6 to 2.6.7 [`#9`](https://github.com/scriptex/gitlab-calendar/pull/9)
- Update dependency release-it to v14.12.1 [`#7`](https://github.com/scriptex/gitlab-calendar/pull/7)
- Update dependency date-fns to v2.28.0 [`#6`](https://github.com/scriptex/gitlab-calendar/pull/6)
- [ImgBot] Optimize images [`#5`](https://github.com/scriptex/gitlab-calendar/pull/5)
- Update dependency release-it to v15.1.1 [`20b4b1e`](https://github.com/scriptex/gitlab-calendar/commit/20b4b1e20a8f948550d8c506fac2a9d1a336224a)
- Update dependency release-it to v14.14.3 [`0ce7133`](https://github.com/scriptex/gitlab-calendar/commit/0ce713368b6ad1999f20104ce0d2005367d51541)
- Update dependency release-it to v14.14.1 [`6d24749`](https://github.com/scriptex/gitlab-calendar/commit/6d2474954d13f503b11300b7f6f06201d7948613)

#### 1.0.0

> 6 December 2021

- Add documentation [`#4`](https://github.com/scriptex/gitlab-calendar/pull/4)
- Initial commit [`daccf85`](https://github.com/scriptex/gitlab-calendar/commit/daccf855aa9ea9ef2685c2c9de0e7edab5e53335)
- Release 1.0.0 [`741e23a`](https://github.com/scriptex/gitlab-calendar/commit/741e23a7eb9e44444d1638c8959aefd86008bf16)


src/gitlab-calendar.ts
import { format } from 'date-fns';
import { select, Selection } from 'd3-selection';

export type ActivityCalendarSVG = Selection<SVGSVGElement, unknown, null, undefined>;

export type ActivityCalendarMonth = {
	x: number;
	month: number;
};

export type ActivityCalendarGroup = {
	day: number;
	date: Date;
	count: number;
};

export type ActivityCalendarDay = {
	y: number;
	text: string;
};

export type ActivityCalendarWeekday = 'monday' | 'wednesday' | 'friday' | 'saturday' | 'sunday';

export type ActivityCalendarLegendValue = {
	min: number;
	title: string;
};

export type GitlabActivityCalendarOptions = {
	daySize: number;
	hintText: string;
	daySpace: number;
	utcOffset: number;
	dayTitles: Record<ActivityCalendarWeekday, string>;
	monthsAgo: number;
	monthNames: string[];
	inputFormat: string;
	weekdayNames: string[];
	legendValues: ActivityCalendarLegendValue[];
	firstDayOfWeek: number;
	tooltipDateFormat: string;
	tooltipFormatter: (count: number, dayName: string, dateText: string) => string;
};

export class GitlabActivityCalendar {
	private _daySizeWithSpace = 0;
	private _millisecondsPerDay = 1000 * 60 * 60 * 24;

	private _defaultOptions: GitlabActivityCalendarOptions = {
		daySize: 15,
		hintText: 'Issues, merge requests, pushes, and comments.',
		daySpace: 1,
		utcOffset: 0,
		dayTitles: {
			monday: 'M',
			wednesday: 'W',
			friday: 'F',
			saturday: 'S',
			sunday: 'S'
		},
		monthsAgo: 12,
		monthNames: ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'],
		inputFormat: 'YYYY-MM-DD',
		weekdayNames: ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'],
		legendValues: [
			{ title: 'No contributions', min: 0 },
			{ title: '1-9 contributions', min: 1 },
			{ title: '10-19 contributions', min: 10 },
			{ title: '20-29 contributions', min: 20 },
			{ title: '30+ contributions', min: 30 }
		],
		firstDayOfWeek: 0,
		tooltipDateFormat: 'MMM d, YYYY',
		tooltipFormatter: (count: number, dayName: string, dateText: string) => {
			let contribText = 'No contributions';

			if (count > 0) {
				contribText = count === 1 ? '1 contribution' : `${count} contributions`;
			}

			return `${contribText} on ${dayName} ${dateText}`;
		}
	};

	private svg: ActivityCalendarSVG;

	private months: ActivityCalendarMonth[] = [];
	private timestamps: Array<ActivityCalendarGroup[]> = [];

	public options: GitlabActivityCalendarOptions = this._defaultOptions;

	constructor(container: HTMLElement, data: Record<string, number>, options: Partial<GitlabActivityCalendarOptions>) {
		this.options = {
			...this._defaultOptions,
			...options
		};

		const group = this.buildArrays(data);

		this._daySizeWithSpace = this.options.daySize + this.options.daySpace * 2;

		this.svg = this.renderSvg(container, group);

		this.renderDays();
		this.renderMonths();
		this.renderDayTitles();
		this.renderKey();
		this.renderHint();
		this.addTitles();
	}

	private buildArrays(data: Record<string, number>): number {
		let group = 0;

		const today = this.getSystemDate(this.options.utcOffset);

		today.setHours(0, 0, 0, 0);

		const timeAgo = new Date(today);

		timeAgo.setMonth(today.getMonth() - this.options.monthsAgo);

		const days = this.getDayDifference(timeAgo, today);

		for (let i = 0; i <= days; i += 1) {
			const date = new Date(timeAgo);

			date.setDate(date.getDate() + i);

			const day = date.getDay();

			const count = data[format(date, this.options.inputFormat)] || 0;

			if ((day === this.options.firstDayOfWeek && i !== 0) || i === 0) {
				this.timestamps.push([]);
				group += 1;
			}

			const innerArray = this.timestamps[group - 1];

			innerArray.push({ count, date, day });
		}

		return group;
	}

	private getDayName = (date: Date): string => this.options.weekdayNames[date.getDay()];

	private getDayDifference = (a: Date, b: Date): number => {
		const date1 = Date.UTC(a.getFullYear(), a.getMonth(), a.getDate());
		const date2 = Date.UTC(b.getFullYear(), b.getMonth(), b.getDate());

		return Math.floor((date2 - date1) / this._millisecondsPerDay);
	};

	private getSystemDate = (systemUtcOffsetSeconds: number): Date => {
		const d = new Date();
		const localUtcOffsetMinutes = 0 - d.getTimezoneOffset();
		const systemUtcOffsetMinutes = systemUtcOffsetSeconds / 60;

		d.setMinutes(d.getMinutes() - localUtcOffsetMinutes + systemUtcOffsetMinutes);

		return d;
	};

	private formatTooltipText = ({ date, count }: { date: Date; count: number }): string => {
		const d = new Date(date);
		const dayName = this.getDayName(d);
		const dateText = format(d, this.options.tooltipDateFormat);

		return this.options.tooltipFormatter(count, dayName, dateText);
	};

	private getLevelFromContributions = (count: number) => {
		if (count <= 0) {
			return 0;
		}

		const nextLevel = this.options.legendValues.findIndex(({ min }) => count < min);

		return nextLevel >= 0 ? nextLevel - 1 : this.options.legendValues.length - 1;
	};

	private getExtraWidthPadding(group: number): number {
		let extraWidthPadding = 0;

		const lastColMonth = this.timestamps[group - 1][0].date.getMonth();
		const secondLastColMonth = this.timestamps[group - 2][0].date.getMonth();

		if (lastColMonth !== secondLastColMonth) {
			extraWidthPadding = 6;
		}

		return extraWidthPadding;
	}

	private renderSvg(container: HTMLElement, group: number): ActivityCalendarSVG {
		const width = (group + 1) * this._daySizeWithSpace + this.getExtraWidthPadding(group);

		return select(container).append('svg').attr('width', width).attr('height', 167);
	}

	private dayYPos(day: number): number {
		return this._daySizeWithSpace * ((day + 7 - this.options.firstDayOfWeek) % 7);
	}

	private renderDays(): void {
		this.svg
			.selectAll('g')
			.data(this.timestamps)
			.enter()
			.append('g')
			.attr('transform', (group: ActivityCalendarGroup[], i: number) => {
				group.forEach((stamp: ActivityCalendarGroup, a: number) => {
					if (a === 0 && stamp.day === this.options.firstDayOfWeek) {
						const month = stamp.date.getMonth();
						const x = this._daySizeWithSpace * i + 1 + this._daySizeWithSpace;
						const lastMonth = this.months[this.months.length - 1];

						if (
							lastMonth == null ||
							(month !== lastMonth.month && x - this._daySizeWithSpace !== lastMonth.x)
						) {
							this.months.push({ month, x });
						}
					}
				});

				return `translate(${this._daySizeWithSpace * i + 1 + this._daySizeWithSpace}, 18)`;
			})
			.selectAll('rect')
			.data((stamp: ActivityCalendarGroup[]) => stamp)
			.enter()
			.append('rect')
			.attr('x', '0')
			.attr('y', (stamp: ActivityCalendarGroup) => this.dayYPos(stamp.day))
			.attr('width', this.options.daySize)
			.attr('height', this.options.daySize)
			.attr('data-level', (stamp: ActivityCalendarGroup) => this.getLevelFromContributions(stamp.count))
			.attr('title', (stamp: ActivityCalendarGroup) => this.formatTooltipText(stamp));
	}

	private renderDayTitles(): void {
		const firstDayOfWeekChoices: Record<'sunday' | 'monday' | 'saturday', number> = {
			sunday: 0,
			monday: 1,
			saturday: 6
		};

		const days: ActivityCalendarDay[] = [
			{
				text: this.options.dayTitles.monday,
				y: 29 + this.dayYPos(1)
			},
			{
				text: this.options.dayTitles.wednesday,
				y: 29 + this.dayYPos(3)
			},
			{
				text: this.options.dayTitles.friday,
				y: 29 + this.dayYPos(5)
			}
		];

		if (this.options.firstDayOfWeek === firstDayOfWeekChoices.monday) {
			days.push({
				text: this.options.dayTitles.sunday,
				y: 29 + this.dayYPos(7)
			});
		} else if (this.options.firstDayOfWeek === firstDayOfWeekChoices.saturday) {
			days.push({
				text: this.options.dayTitles.saturday,
				y: 29 + this.dayYPos(6)
			});
		}

		this.svg
			.append('g')
			.selectAll('text')
			.data(days)
			.enter()
			.append('text')
			.attr('text-anchor', 'middle')
			.attr('x', 8)
			.attr('y', (day: ActivityCalendarDay) => day.y)
			.text((day: ActivityCalendarDay) => day.text);
	}

	private renderMonths(): void {
		this.svg
			.append('g')
			.selectAll('text')
			.data(this.months)
			.enter()
			.append('text')
			.attr('x', (date: ActivityCalendarMonth) => date.x)
			.attr('y', 10)
			.text((date: ActivityCalendarMonth) => this.options.monthNames[date.month]);
	}

	private renderKey(): void {
		this.svg
			.append('g')
			.attr('transform', `translate(18, ${this._daySizeWithSpace * 8 + 16})`)
			.selectAll('rect')
			.data(this.options.legendValues)
			.enter()
			.append('rect')
			.attr('width', this.options.daySize)
			.attr('height', this.options.daySize)
			.attr('x', (_, i) => this._daySizeWithSpace * i)
			.attr('y', 0)
			.attr('data-level', (_, i) => i)
			.attr('title', x => x.title);
	}

	private renderHint(): void {
		this.svg
			.append('g')
			.attr(
				'transform',
				`translate(${this.svg.property('width').baseVal.value}, ${this._daySizeWithSpace * 8 + 25})`
			)
			.append('text')
			.attr('text-anchor', 'end')
			.text(this.options.hintText);
	}

	private addTitles(): void {
		this.svg.selectAll('rect').each(function () {
			const rect: SVGRectElement | null = this as SVGRectElement | null;

			if (!rect) {
				return;
			}

			const group = document.createElementNS('http://www.w3.org/2000/svg', 'g');
			const titleString = rect.getAttribute('title') || '';
			const titleElement = document.createElementNS('http://www.w3.org/2000/svg', 'title');

			rect?.parentNode?.insertBefore(group, rect);

			titleElement.innerHTML = titleString;

			group.appendChild(titleElement);
			group.appendChild(rect);
		});
	}
}

export default GitlabActivityCalendar;


src/index.ts
import format from 'date-fns/format';
import { select, Selection } from 'd3-selection';

export type ActivityCalendarSVG = Selection<SVGSVGElement, unknown, null, undefined>;

export type ActivityCalendarMonth = {
	x: number;
	month: number;
};

export type ActivityCalendarGroup = {
	day: number;
	date: Date;
	count: number;
};

export type ActivityCalendarDay = {
	y: number;
	text: string;
};

export type ActivityCalendarWeekday = 'monday' | 'wednesday' | 'friday' | 'saturday' | 'sunday';

export type ActivityCalendarLegendValue = {
	min: number;
	title: string;
};

export type GitlabCalendarOptions = {
	daySize: number;
	hintText: string;
	daySpace: number;
	utcOffset: number;
	dayTitles: Record<ActivityCalendarWeekday, string>;
	monthsAgo: number;
	monthNames: string[];
	inputFormat: string;
	weekdayNames: string[];
	legendValues: ActivityCalendarLegendValue[];
	firstDayOfWeek: number;
	tooltipDateFormat: string;
	tooltipFormatter: (count: number, dayName: string, dateText: string) => string;
};

export class GitlabCalendar {
	private _daySizeWithSpace = 0;
	private _millisecondsPerDay = 1000 * 60 * 60 * 24;

	private _defaultOptions: GitlabCalendarOptions = {
		daySize: 15,
		hintText: 'Issues, merge requests, pushes, and comments.',
		daySpace: 1,
		utcOffset: 0,
		dayTitles: {
			monday: 'M',
			wednesday: 'W',
			friday: 'F',
			saturday: 'S',
			sunday: 'S'
		},
		monthsAgo: 12,
		monthNames: ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'],
		inputFormat: 'yyyy-MM-dd',
		weekdayNames: ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'],
		legendValues: [
			{ title: 'No contributions', min: 0 },
			{ title: '1-9 contributions', min: 1 },
			{ title: '10-19 contributions', min: 10 },
			{ title: '20-29 contributions', min: 20 },
			{ title: '30+ contributions', min: 30 }
		],
		firstDayOfWeek: 0,
		tooltipDateFormat: 'MMM d, yyyy',
		tooltipFormatter: (count: number, dayName: string, dateText: string) => {
			let contribText = 'No contributions';

			if (count > 0) {
				contribText = count === 1 ? '1 contribution' : `${count} contributions`;
			}

			return `${contribText} on ${dayName} ${dateText}`;
		}
	};

	private svg: ActivityCalendarSVG;

	private months: ActivityCalendarMonth[] = [];
	private timestamps: Array<ActivityCalendarGroup[]> = [];

	public options: GitlabCalendarOptions = this._defaultOptions;

	constructor(container: HTMLElement, data: Record<string, number>, options: Partial<GitlabCalendarOptions>) {
		this.options = {
			...this._defaultOptions,
			...options
		};

		const group = this.buildArrays(data);

		this._daySizeWithSpace = this.options.daySize + this.options.daySpace * 2;

		this.svg = this.renderSvg(container, group);

		this.renderDays();
		this.renderMonths();
		this.renderDayTitles();
		this.renderKey();
		this.renderHint();
		this.addTitles();
	}

	private buildArrays(data: Record<string, number>): number {
		let group = 0;

		const today = this.getSystemDate(this.options.utcOffset);

		today.setHours(0, 0, 0, 0);

		const timeAgo = new Date(today);

		timeAgo.setMonth(today.getMonth() - this.options.monthsAgo);

		const days = this.getDayDifference(timeAgo, today);

		for (let i = 0; i <= days; i += 1) {
			const date = new Date(timeAgo);

			date.setDate(date.getDate() + i);

			const day = date.getDay();

			const count = data[format(date, this.options.inputFormat)] || 0;

			if ((day === this.options.firstDayOfWeek && i !== 0) || i === 0) {
				this.timestamps.push([]);
				group += 1;
			}

			const innerArray = this.timestamps[group - 1];

			innerArray.push({ count, date, day });
		}

		return group;
	}

	private getDayName = (date: Date): string => this.options.weekdayNames[date.getDay()];

	private getDayDifference = (a: Date, b: Date): number => {
		const date1 = Date.UTC(a.getFullYear(), a.getMonth(), a.getDate());
		const date2 = Date.UTC(b.getFullYear(), b.getMonth(), b.getDate());

		return Math.floor((date2 - date1) / this._millisecondsPerDay);
	};

	private getSystemDate = (systemUtcOffsetSeconds: number): Date => {
		const d = new Date();
		const localUtcOffsetMinutes = 0 - d.getTimezoneOffset();
		const systemUtcOffsetMinutes = systemUtcOffsetSeconds / 60;

		d.setMinutes(d.getMinutes() - localUtcOffsetMinutes + systemUtcOffsetMinutes);

		return d;
	};

	private formatTooltipText = ({ date, count }: { date: Date; count: number }): string => {
		const d = new Date(date);
		const dayName = this.getDayName(d);
		const dateText = format(d, this.options.tooltipDateFormat);

		return this.options.tooltipFormatter(count, dayName, dateText);
	};

	private getLevelFromContributions = (count: number) => {
		if (count <= 0) {
			return 0;
		}

		const nextLevel = this.options.legendValues.findIndex(({ min }) => count < min);

		return nextLevel >= 0 ? nextLevel - 1 : this.options.legendValues.length - 1;
	};

	private getExtraWidthPadding(group: number): number {
		let extraWidthPadding = 0;

		const lastColMonth = this.timestamps[group - 1][0].date.getMonth();
		const secondLastColMonth = this.timestamps[group - 2][0].date.getMonth();

		if (lastColMonth !== secondLastColMonth) {
			extraWidthPadding = 6;
		}

		return extraWidthPadding;
	}

	private renderSvg(container: HTMLElement, group: number): ActivityCalendarSVG {
		const width = (group + 1) * this._daySizeWithSpace + this.getExtraWidthPadding(group);

		return select(container).append('svg').attr('width', width).attr('height', 167);
	}

	private dayYPos(day: number): number {
		return this._daySizeWithSpace * ((day + 7 - this.options.firstDayOfWeek) % 7);
	}

	private renderDays(): void {
		this.svg
			.selectAll('g')
			.data(this.timestamps)
			.enter()
			.append('g')
			.attr('transform', (group: ActivityCalendarGroup[], i: number) => {
				group.forEach((stamp: ActivityCalendarGroup, a: number) => {
					if (a === 0 && stamp.day === this.options.firstDayOfWeek) {
						const month = stamp.date.getMonth();
						const x = this._daySizeWithSpace * i + 1 + this._daySizeWithSpace;
						const lastMonth = this.months[this.months.length - 1];

						if (
							lastMonth == null ||
							(month !== lastMonth.month && x - this._daySizeWithSpace !== lastMonth.x)
						) {
							this.months.push({ month, x });
						}
					}
				});

				return `translate(${this._daySizeWithSpace * i + 1 + this._daySizeWithSpace}, 18)`;
			})
			.selectAll('rect')
			.data((stamp: ActivityCalendarGroup[]) => stamp)
			.enter()
			.append('rect')
			.attr('x', '0')
			.attr('y', (stamp: ActivityCalendarGroup) => this.dayYPos(stamp.day))
			.attr('width', this.options.daySize)
			.attr('height', this.options.daySize)
			.attr('data-level', (stamp: ActivityCalendarGroup) => this.getLevelFromContributions(stamp.count))
			.attr('title', (stamp: ActivityCalendarGroup) => this.formatTooltipText(stamp));
	}

	private renderDayTitles(): void {
		const firstDayOfWeekChoices: Record<'sunday' | 'monday' | 'saturday', number> = {
			sunday: 0,
			monday: 1,
			saturday: 6
		};

		const days: ActivityCalendarDay[] = [
			{
				text: this.options.dayTitles.monday,
				y: 29 + this.dayYPos(1)
			},
			{
				text: this.options.dayTitles.wednesday,
				y: 29 + this.dayYPos(3)
			},
			{
				text: this.options.dayTitles.friday,
				y: 29 + this.dayYPos(5)
			}
		];

		if (this.options.firstDayOfWeek === firstDayOfWeekChoices.monday) {
			days.push({
				text: this.options.dayTitles.sunday,
				y: 29 + this.dayYPos(7)
			});
		} else if (this.options.firstDayOfWeek === firstDayOfWeekChoices.saturday) {
			days.push({
				text: this.options.dayTitles.saturday,
				y: 29 + this.dayYPos(6)
			});
		}

		this.svg
			.append('g')
			.selectAll('text')
			.data(days)
			.enter()
			.append('text')
			.attr('text-anchor', 'middle')
			.attr('x', 8)
			.attr('y', (day: ActivityCalendarDay) => day.y)
			.text((day: ActivityCalendarDay) => day.text);
	}

	private renderMonths(): void {
		this.svg
			.append('g')
			.selectAll('text')
			.data(this.months)
			.enter()
			.append('text')
			.attr('x', (date: ActivityCalendarMonth) => date.x)
			.attr('y', 10)
			.text((date: ActivityCalendarMonth) => this.options.monthNames[date.month]);
	}

	private renderKey(): void {
		this.svg
			.append('g')
			.attr('transform', `translate(18, ${this._daySizeWithSpace * 8 + 16})`)
			.selectAll('rect')
			.data(this.options.legendValues)
			.enter()
			.append('rect')
			.attr('width', this.options.daySize)
			.attr('height', this.options.daySize)
			.attr('x', (_, i) => this._daySizeWithSpace * i)
			.attr('y', 0)
			.attr('data-level', (_, i) => i)
			.attr('title', x => x.title);
	}

	private renderHint(): void {
		this.svg
			.append('g')
			.attr(
				'transform',
				`translate(${this.svg.property('width').baseVal.value}, ${this._daySizeWithSpace * 8 + 25})`
			)
			.append('text')
			.attr('text-anchor', 'end')
			.text(this.options.hintText);
	}

	private addTitles(): void {
		this.svg.selectAll('rect').each(function () {
			const rect: SVGRectElement | null = this as SVGRectElement | null;

			if (!rect) {
				return;
			}

			const group = document.createElementNS('http://www.w3.org/2000/svg', 'g');
			const titleString = rect.getAttribute('title') || '';
			const titleElement = document.createElementNS('http://www.w3.org/2000/svg', 'title');

			rect.parentNode?.insertBefore(group, rect);

			titleElement.innerHTML = titleString;

			group.appendChild(titleElement);
			group.appendChild(rect);
		});
	}
}

export default GitlabCalendar;
